dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,245 @@
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
+    if types._is_torch_tensor(sample):
+        if sample.device.type != 'cpu':
+            raise TypeError("GPU tensors are not supported")
+        return sample.numpy()
+    elif isinstance(sample, tensors.TensorCPU):
+        return np.array(sample)
+    raise TypeError(
+        "Unsupported callback return type. Expected numpy array, pytorch or mxnet cpu tensors, "
+        "or list or tuple of them.")
+
+
+def _apply_to_sample(func, sample, *args):
+    """Apply to a sample traversing the nesting of the data (tuple/list).
+
+    Parameters
+    ----------
+    func : callable
+        Function to be applied to every sample data object
+    sample : sample object or any nesting of those in tuple/list
+        Representation of sample
+    """
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(_apply_to_sample(func, part, *args) for part in sample)
+    else:
+        # we unpacked all nesting levels, now is actual data:
+        return func(sample, *args)
+
+
+class SharedBatchWriter:
+    """SharedBatchWriter can serialize and write batch into given shared
+    memory chunk (``mem_batch``).
+    """
+
+    def __init__(self, mem_batch: SharedMemChunk, batch):
+        import_numpy()
+        self.mem_batch = mem_batch
+        self.data_size = 0
+        self.meta_data_size = 0
+        self._write_batch(batch)
+
+    def _prepare_samples_meta(self, indexed_samples):
+        """Calculate metadata and total size of data to be serialized"""
+        data_size = 0
+
+        def make_meta(np_array):
+            nonlocal data_size
+            offset = data_size
+            data_size += np_array.nbytes
+            return SampleMeta(offset, np_array.shape, np_array.dtype, np_array.nbytes)
+
+        meta = []
+        for idx, sample in indexed_samples:
+            meta.append((idx, _apply_to_sample(make_meta, sample)))
+        return meta, data_size
+
+    def _add_array_to_batch(self, np_array, memview):
+        sample_size = np_array.nbytes
+        offset = self.written_size
+        self.written_size += sample_size
+        buffer = memview[offset:(offset + sample_size)]
+        shared_array = np.ndarray(
+            np_array.shape, dtype=np_array.dtype, buffer=buffer)
+        shared_array[:] = np_array[:]
+
+    def _write_batch(self, batch):
+        if not batch:
+            return
+        batch = [(idx, _apply_to_sample(_to_numpy, sample)) for idx, sample in batch]
+        meta, data_size = self._prepare_samples_meta(batch)
+        serialized_meta = pickle.dumps(meta)
+        self.meta_data_size = len(serialized_meta)
+        self.data_size = data_size
+        needed_capacity = self.data_size + self.meta_data_size
+        if self.mem_batch.capacity < needed_capacity:
+            self.mem_batch.resize(max([needed_capacity, 2 * self.mem_batch.capacity]))
```

Comment:
nitpick - but I think it reads better
```suggestion
            self.mem_batch.resize(max(needed_capacity, 2 * self.mem_batch.capacity))
```

Reply:
I agree.